asm/osInvalDCache.s

```diff
@@ -8,54 +9,88 @@
-/* 006E74 80006274 0109082B */  sltu  $at, $t0, $t1
-/* 006E78 80006278 1420FFFD */  bnez  $at, .L80006270
-/* 006E7C 8000627C 25080010 */   addiu $t0, $t0, 0x10
-.L80006280:
-/* 006E80 80006280 03E00008 */  jr    $ra
-/* 006E84 80006284 00000000 */   nop   
 
-.L80006288:
-/* 006E88 80006288 3C088000 */  lui   $t0, 0x8000
-/* 006E8C 8000628C 010B4821 */  addu  $t1, $t0, $t3
-/* 006E90 80006290 2529FFF0 */  addiu $t1, $t1, -0x10
-.L80006294:
-/* 006E94 80006294 BD010000 */  cache 1, ($t0)
-/* 006E98 80006298 0109082B */  sltu  $at, $t0, $t1
-/* 006E9C 8000629C 1420FFFD */  bnez  $at, .L80006294
-/* 006EA0 800062A0 25080010 */   addiu $t0, 0x10
-/* 006EA4 800062A4 03E00008 */  jr    $ra
-/* 006EA8 800062A8 00000000 */   nop   
+/**
+ *  void osInvalDCache(void* vaddr, s32 nbytes);
+ *
+ *  Invalidates the CPU Data Cache for `nbytes` at `vaddr`.
+ *  The cache is not automatically synced with physical memory, so cache
+ *  lines must be invalidated to ensure old data is not used in place of
+ *  newly available data supplied by an external agent in a DMA operation.
+ *
+ *  If `vaddr` is not aligned to a cache line boundary, or nbytes is not a
+ *  multiple of the data cache line size (16 bytes) a larger region is
+ *  invalidated.
+ *
+ *  If the amount to invalidate is greater than the data cache size, 8192,
+ *  the entire data cache is invalidated.
+ */
+LEAF(osInvalDCache)
+    # If the amount to invalidate is less or equal to 0, return immediately
+    blez    $a1, .ret
+     nop
+    # If the amount to invalidate is as large as or larger than
+    # the data cache size, invalidate all
+    li      $t3, DCACHE_SIZE
+    sltu    $at, $a1, $t3
+    beqz    $at, .all
+     nop
+    # Ensure end address doesn't wrap around and end up smaller
+    # than the start address
+    move    $t0, $a0
+    addu    $t1, $a0, $a1
+    sltu    $at, $t0, $t1
+    beqz    $at, .ret
+     nop
+    # Mask start with cache line
+    andi    $t2, $t0, DCACHE_LINEMASK
+    # If mask is not zero, the start is not cache aligned
+    beqz    $t2, .aligned_start
+     addiu  $t1, $t1, -DCACHE_LINESIZE
+    # Subtract mask result to align to cache line
+    subu    $t0, $t0, $t2
+    # Hit-Writeback-Invalidate unaligned part
+    cache   (CACH_PD | C_HWBINV), ($t0)
+    sltu    $at, $t0, $t1
+    # If that's all there is to do, return early
+    beqz    $at, .ret
+     nop
+    addiu   $t0, $t0, DCACHE_LINESIZE
+.aligned_start:
+    # Mask end with cache line
+    andi    $t2, $t1, DCACHE_LINEMASK
+    # If mask is not zero, the end is not cache aligned
+    beqz    $t2, 1f
+     nop
+    # Subtract mask result to align to cache line
+    subu    $t1, $t1, $t2
+    # Hit-Writeback-Invalidate unaligned part
+    cache   (CACH_PD | C_HWBINV), DCACHE_LINESIZE($t1)
+    sltu    $at, $t1, $t0
+    # If that's all there is to do, return early
+    bnez    $at, .ret
+     nop
+    # Invalidate the rest
+1:
+    # Index-Invalidate
+    cache   (CACH_PD | C_HINV), ($t0)
```

Comment:
```suggestion
    # Hit-Invalidate
    cache   (CACH_PD | C_HINV), ($t0)
```
the cache operations usage here looked like a complete mess at first but actually makes sense